core.js

```diff
@@ -60,16 +44,92 @@ function _check(buffer, headers, options) {
 }
 
 export async function fileTypeFromTokenizer(tokenizer) {
-	try {
-		return new FileTypeParser().parse(tokenizer);
-	} catch (error) {
-		if (!(error instanceof strtok3.EndOfStreamError)) {
-			throw error;
-		}
+	await new FileTypeParser().fileTypeFromTokenizer(tokenizer);
+}
+
+export class TokenizerPositionError extends Error {
+	constructor(message = 'Tokenizer position changed during operation. This is not allowed as it makes following detectors read from the wrong offset. If the offset change is intentional, use recursion.') {
+		super(message);
+		this.name = 'TokenizerPositionError';
 	}
 }
 
-class FileTypeParser {
+export class FileTypeParser {
+	/**
+	 * The detectionImpossible below can be returned by custom detectors
+	 * when they are certain that the file type can not be determined,
+	 * neither by the detector itself, nor by any further detector.
+	 * This will interrupt the file type detection.
+	 */
+	static detectionImpossible = 'detection_impossible';
+
+	constructor(options) {
+		this.detectors = options?.customDetectors;
+
+		this.runCustomDetectors = this.runCustomDetectors.bind(this);
+		this.fileTypeFromTokenizer = this.fileTypeFromTokenizer.bind(this);
+		this.fileTypeFromBuffer = this.fileTypeFromBuffer.bind(this);
+		this.parse = this.parse.bind(this);
+	}
+
+	async runCustomDetectors(tokenizer) {
+		const {detectors} = this;
+		const initialPosition = tokenizer?.position;
+
+		for (const detector of detectors || []) {
+			const fileType = await detector(tokenizer);
+			if (fileType) {
+				return fileType;
+			}
+
+			if (initialPosition !== tokenizer?.position) {
+				throw new TokenizerPositionError();
+			}
+		}
+
+		return undefined;
+	}
+
+	async fileTypeFromTokenizer(tokenizer) {
+		const {parse, runCustomDetectors} = this;
+		try {
+			const customFileType = await runCustomDetectors(tokenizer);
+			return (customFileType || await parse(tokenizer));
+		} catch (error) {
+			if (!(error instanceof strtok3.EndOfStreamError)) {
+				throw error;
+			}
+		}
+	}
+
+	async fileTypeFromBuffer(input) {
+		if (!(input instanceof Uint8Array || input instanceof ArrayBuffer)) {
+			throw new TypeError(`Expected the \`input\` argument to be of type \`Uint8Array\` or \`Buffer\` or \`ArrayBuffer\`, got \`${typeof input}\``);
+		}
+
+		const buffer = input instanceof Uint8Array ? input : new Uint8Array(input);
+
+		if (!(buffer?.length > 1)) {
+			return;
+		}
+
+		return this.fileTypeFromTokenizer(strtok3.fromBuffer(buffer));
+	}
+
+	async fileTypeFromBlob(blob) {
+		const buffer = await blob.arrayBuffer();
+		return this.fileTypeFromBuffer(new Uint8Array(buffer));
+	}
+
+	async fileTypeFromStream(stream) {
```

Comment:
Inconsistent with function

Reply:
I don't see what you mean, I'm afraid. Do you refer to `this.fileTypeFromTokenizer(tokenizer);` being called instead of `fileTypeFromTokenizer(tokenizer);`?

Reply:
My bad, my comment was applicable to the last line only:

```js
async fileTypeFromStream(stream) {
```
which is different then:
```
fileTypeStream(readableStream, {sampleSize = minimumBytes} = {}, fileTypeOptions)
```
the sample size was probable already missing.

So it's essentially inconsistent with TypeScript declaration, as reported in https://github.com/sindresorhus/file-type/pull/603#discussion_r1367942822

Reply:
Aren't `fromStream(...)` and `stream(...)` methods getting mixed up here? (`fileTypeFromStream(...)` and `fileTypeStream(...)` respectively)